pallets/liquidity-pools-gateway/queue/src/benchmarking.rs

```diff
@@ -0,0 +1,64 @@
+// Copyright 2024 Centrifuge Foundation (centrifuge.io).
+// This file is part of Centrifuge chain project.
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use frame_benchmarking::{account, impl_benchmark_test_suite, v2::*};
+use frame_system::RawOrigin;
+use parity_scale_codec::EncodeLike;
+
+use super::*;
+
+#[benchmarks(
+	where
+		T: Config,
+		T::AccountId: EncodeLike<<T as frame_system::Config>::AccountId>,
+)]
+mod benchmarks {
+	use super::*;
+
+	#[benchmark]
+	fn process_message() -> Result<(), BenchmarkError> {
+		let caller: T::AccountId = account("acc_0", 0, 0);
+		let message = T::Message::default();
```

Comment:
That will be tricky since the message itself is irrelevant given that it gets passed to a mock.

Reply:
> it gets passed to a mock

Note that when the benchmark is executed for the real runtimes, the mock is no longer used.

@wischli we already talked about adding a PoV for the MaxEncodedLen to the resulting weights

Reply:
> Note that when the benchmark is executed for the real runtimes, the mock is no longer used.

Oh, that explains a lot then. Where does this magical switch from mock runtime to actual runtime happen?

Yes, we should definitely add that there before merging this.

Reply:
The magic happens here: https://github.com/centrifuge/centrifuge-chain/blob/main/runtime/centrifuge/src/lib.rs#L2775

That macro calls the benchmarks of those pallets using the real runtime instead of the mock. That's why sometimes the benchmarking is so difficult to achieve. You should be able to create a benchmark that works fine for any runtime configuration.

Right now, the maximum message is no more than 256 bytes. So we're "safe", maybe just adding to the extrinsic is enough:

#[pallet::weight(T::WeightInfo::process_message().saturating_add(Weights::from_parts(0, T::Message::max_encoded_len())))]

I would not go with a huge change because the batch process will need some special treatment regarding this. I'll do it in the batch PR once I rebase over this.

Reply:
Thank you for the pointer, I'll expand that macro and take a look.

As for the `process_message` weight, with the changes that I'm currently working on, this will be a bit easier since that extrinsic will now submit the message to the queue. It will be then processed `on_idle` and return a `DispatchResultWithPostInfo` that can have more granular weights, as per [this](https://github.com/centrifuge/centrifuge-chain/pull/1937/files#diff-74702c928dc1b5f82b0ba7573573ef25bc263370ba55e341eeeadc7dec1ef255R567).

Reply:
If you're able to return directly the weight there, then I would say we're fine

Reply:
BTW (NIT), if you're only interested in the `weight`, maybe it's interesting to return a tuple `(DispatchResult, Weight)` instead. That way, you can get rid of the unused `pay_fees`, and you have the `weight` for both successful processing and failures. I was experimenting with that in the batch PR; maybe it's useful in that part too.

Reply:
Awesome idea, it should make our lives easier.